library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Sum.hs

```diff
@@ -0,0 +1,224 @@
+
+lookupType :: Types.LookupType m => PassTypes.SumName -> m Type.Type
+lookupType sumName = do
+  getSumDesc sumName >>= pure . fst
+
+oneArgFunctionType :: Type.Type -> Type.Type -> Type.Type
+oneArgFunctionType arg result =
+  Types.pointerOf
+    Type.FunctionType
+      { Type.resultType = result,
+        Type.argumentTypes = [arg],
+        Type.isVarArg = False
+      }
+
+makeCase ::
+  Types.Define m =>
+  PassTypes.SumName ->
+  Type.Type ->
+  [Type.Type] ->
+  AST.Operand ->
+  [AST.Operand] ->
+  [AST.Operand] ->
+  m AST.Operand
+makeCase sumName outputType caseTypes term cases environments = do
+  (sumType, variantDescs) <- getSumDesc sumName
+  let values = map tagConstant [0 .. length variantDescs - 1]
+  let variantNames = map fst variantDescs
+  let variantTypes = map snd variantDescs
+  let expectedCaseTypes = map (`oneArgFunctionType` outputType) variantTypes
+  if caseTypes /= expectedCaseTypes
+    then
+      throw @"err" $
+        Types.MismatchedCaseTypes $
+          "typechecker allowed mismatched case types: expected "
+            <> show expectedCaseTypes
+            <> "; got "
+            <> show caseTypes
+            <> "; outputType "
+            <> show outputType
+            <> "; variantTypes "
+            <> show variantTypes
+    else pure ()
+  indexPtr <- getIndexPtr term
+  index <- Block.load indexType indexPtr
+  variantPtrLoc <- getVariantPtr term
+  variantPtr <- Block.load variantPtrType variantPtrLoc
+  let appliedCases =
+        map
+          ( \(ty, (caseFunc, environment)) -> do
+              castedPtr <- Block.bitCast variantPtr (Types.pointerOf ty)
+              variant <- Block.load ty castedPtr
+              Block.call outputType caseFunc [(environment, []), (variant, [])]
+          )
+          (zip variantTypes $ zip cases environments)
+  Block.generateSwitch outputType index variantNames values appliedCases
+
+-- | Given the name of a sum type and a compiled variant term,
+-- | allocate a sum type and store the given term in it.
+-- |
+-- | WARNING:  The memory allocations herein leak memory, pending our
+-- | implementing garbage collection.
+makeSum ::
+  Types.Define m =>
+  PassTypes.SumName ->
+  PassTypes.VariantName ->
+  Type.Type ->
+  Type.Type ->
+  AST.Operand ->
+  m AST.Operand
+makeSum sumName variantName compiledSumType compiledVariantType variantTerm = do
+  (sumType, variantDescs) <- getSumDesc sumName
+  if compiledSumType /= sumType
+    then
+      throw @"err" $
+        Types.MismatchedSumTypes $
+          "typechecker allowed mismatched sum types: expected "
+            <> show sumType
+            <> "; got "
+            <> show compiledSumType
+    else pure ()
+  variantIndex <- case List.findIndex (\desc -> fst desc == toSymbol variantName) variantDescs of
+    Just index -> pure index
+    Nothing ->
+      throw @"err" $
+        Types.NonExistentVariant "typechecker allowed selection of non-existent variant"
```

Comment:
Seems we might want a hashtable to hold these values instead of lists? Idk though with other semantics, but the search function suggests that to me